File: BlockOfLabs2/laboratory2.4/lab2.4.3.cpp
#include <iostream>
#include <string>
#include <fstream>


using namespace std;


const int CONS_NUM = 1;
const int FILE_NUM = 2;
const int MIN_ARR_SIZE = 2;


void printStatement()
{
	cout << "The program calculates whether the entered\n\t"
		"natural number sequence is increasing.\n\n";
}


// work with way to the file
int pathСheck(bool& isIncorrect)
{
	int path;
	cin >> path;

	if (cin.fail() || cin.get() != '\n')
	{
		cerr << "Invalid numeric input. Try again.\n";
		cin.clear();
		while (cin.get() != '\n');
	}
	else if (path != CONS_NUM && path != FILE_NUM)
		cerr << "Choose only " << CONS_NUM << " or "
		<< FILE_NUM << ". Try again.\n";
	else
		isIncorrect = false;

	return path;
}


int choosingAPath()
{
	cout << "Where will we work through: \n\tConsole: "
		<< CONS_NUM << "\tFile: " << FILE_NUM << "\n\n";

	int path;
	bool isIncorrect = true;
	do
	{
		cout << "Your choice: ";
		path = pathСheck(isIncorrect);
	} while (isIncorrect);

	return path;
}


// block of conditon check
bool isArrIncreasing(double*& arrOfNumb, int arrSize)
{
	for (int i = 1; i < arrSize; i++) {
		if (!(arrOfNumb[i] <= arrOfNumb[i - 1])) {
			return false;
		}
	}

	return true;
}


int resultOfArrChecking(bool isIncrease)
{
	if (isIncrease)
		return 1;
	else
		return 0;
}


/// work in console
int inputArrSize()
{
	bool isIncorrect = true;
	int arrSize;

	do
	{
		cout << "How much number in massive: ";
		cin >> arrSize;

		if (cin.fail() || cin.get() != '\n')
		{
			cerr << "Invalid numeric input. Try again.\n";
			cin.clear();
			while (cin.get() != '\n');
		}
		else if (arrSize < MIN_ARR_SIZE)
			cerr << "Min num is " << MIN_ARR_SIZE << ". Try again.\n";
		else
			isIncorrect = false;
	} while (isIncorrect);

	return arrSize;
}


double enteringTheCurrentNumber(int i)
{
	bool isIncorrect = true;
	double currentNum;

	do
	{
		cout << "Write your " << i + 1 << " number: ";
		cin >> currentNum;

		if (cin.fail() || cin.get() != '\n')
		{
			cerr << "Invalid numeric input. Try again.\n";
			cin.clear();
			while (cin.get() != '\n');
		}
		else
			isIncorrect = false;
	} while (isIncorrect);

	return currentNum;
}


void inputArr(double*& arrOfNumb, int arrSize)
{
	for (int i = 0; i < arrSize; i++)
		arrOfNumb[i] = enteringTheCurrentNumber(i);
}


double viaConsole(double*& arrOfNumb)
{
	int arrSize = inputArrSize();

	inputArr(arrOfNumb, arrSize);

	return arrSize;
}


/// work with file
void fileRestriction()
{
	cout << "\nRules for storing information in a file:\n\t"
		"1.  The first line contains an integer: \n\t\t"
		"the number of array elements;\n\t"
		"2.  The second line is real number\n\t\t"
		"entered separated by spaces.\n\n";
}


void wayCondition(string way, bool& isIncorrect)
{
	if (way.size() > 4)
	{
		string bufstr = way.substr(way.size() - 4);
		if (bufstr == ".txt")
			isIncorrect = false;
		else
			cerr << "Write .txt file.\n";
	}
	else
		cerr << "The path is too short.\n";
}


string inputWayToTheFile()
{
	string way;
	bool isIncorrect = true;

	cout << "Write way to your file: ";
	do
	{
		cin >> way;

		wayCondition(way, isIncorrect);
	} while (isIncorrect);

	return way;
}


bool isFileIntegrity(string fileWay)
{
	bool isIntegrity = false;

	fstream file;
	file.open(fileWay);

	if (file.is_open())
		isIntegrity = true;

	file.clear();
	file.close();

	return isIntegrity;
}


int resultOfReading(bool isCorrect, int arrSize, double*& arrOfNumb)
{
	if (isCorrect) {
		arrOfNumb = NULL;
		return arrSize;
	}
	else {
		cerr << "ERROR in file.";

		return -1;
	}
}


int isCorrectInputFromFile(ifstream& file, int& arrSize, double*& arrOfNumb)
{
	bool isCorrect = true;
	file >> arrSize;
	if (file.fail() || file.get() != '\n')
		isCorrect = false;

	if (arrSize < MIN_ARR_SIZE && isCorrect)
		isCorrect = false;

	if (isCorrect) {
		arrOfNumb = new double[arrSize];
		for (int i = 0; i < arrSize; i++)
		{
			file >> arrOfNumb[i];
			if (file.fail())
				isCorrect = false;
		}
	}

	if (!file.eof() && isCorrect)
		isCorrect = false;

	return resultOfReading(isCorrect, arrSize, arrOfNumb);
}


int isReadingCorrect(string fileWay, int arrSize, double*& arrOfNumb)
{
	ifstream file(fileWay, ios::in);

	int result = isCorrectInputFromFile(file, arrSize, arrOfNumb);

	file.clear();
	file.close();

	return result;
}


int workWithFile(string fileWay, double*& arrOfNumb)
{
	int arrSize = 0;

	int result = isReadingCorrect(fileWay, arrSize, arrOfNumb);

	return result;
}


int workWithIntergrityResoult(bool isIntegrity, string fileWay, double*& arrOfNumb)
{
	if (isIntegrity)
		return workWithFile(fileWay, arrOfNumb);
	else
	{
		cerr << "Bad File.";

		return -1;
	}
}


int viaFile(double*& arrOfNumb)
{
	fileRestriction();

	string fileWay = inputWayToTheFile();
	bool isIntegrity = isFileIntegrity(fileWay);

	return workWithIntergrityResoult(isIntegrity, fileWay, arrOfNumb);
}


/// output console
void outputViaConsole(int result)
{
	if (result)
		cout << "Increase.";
	else
		cout << "Uncreased.";
}


/// output file
string fileCorrectOutput(int result)
{
	if (result)
		return "\nIncrease.";
	else
		return "\nUncreased.";
}


void outputViaFile(int result)
{
	string fileWay = inputWayToTheFile();
	ofstream file(fileWay, ios::app);

	if (file.is_open())
	{
		file << fileCorrectOutput(result);
		cout << "Check your file.";
	}
	else
		cerr << "\nBad output file.";

	file.clear();
	file.close();
}


/// output
void output(int result)
{
	if (result != -1)
	{
		cout << "\n\nYou need to choose where to output the result.\n";
		int option = choosingAPath();

		option == FILE_NUM ? outputViaFile(result) : outputViaConsole(result);
	}
}


/// main block
int main()
{
	printStatement();

	int option = choosingAPath();
	double* arrOfNumb;
	int arrSize = option == FILE_NUM ? viaFile(arrOfNumb) : viaConsole(arrOfNumb);

	if (arrOfNumb != NULL) {
		bool isIncreasing = isArrIncreasing(arrOfNumb, arrSize);
		int result = resultOfArrChecking(isIncreasing);
		output(result);
	}

	delete[] arrOfNumb;
	arrOfNumb = nullptr;
	return 0;
}

File: BlockOfLabs2/laboratory2.4/lab2.4.3.java
package lab2;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.util.Scanner;

class SaveData{
    private boolean isCorrect;
    private int arrSize;

    public SaveData(boolean isCorrect, int arrSize){
        this.isCorrect = isCorrect;
        this.arrSize = arrSize;
    }

    public boolean isCorrect() {
        return isCorrect;
    }

    public void setCorrect(boolean correct) {
        isCorrect = correct;
    }

    public int getArrSize() {
        return arrSize;
    }

    public void setArrSize(int arrSize) {
        this.arrSize = arrSize;
    }
}

public class lab4 {
    static final int CONS_NUM = 1;
    static final int FILE_NUM = 2;
    static final int MIN_ARR_SIZE = 2;

    static void printStatement(){
        System.out.print("""
                         The program calculates whether the entered
                                 natural number sequence is increasing.
                         """);
    }


    // work with way to the file
    static int pathCheck(SaveData SD, Scanner in){
        int path = 0;
        try{
            path = Integer.parseInt(in.nextLine());
        } catch(Exception error){
            System.err.print("Invalid numeric input. Try again.\n");
        }
        if (path != CONS_NUM && path != FILE_NUM)
            System.err.printf("Choose only %d or %d. Try again.\n", CONS_NUM, FILE_NUM);
        else
            SD.setCorrect(false);

        return path;
    }


    static int choosingAPath(Scanner in){
        System.out.printf("""
                Where will we work through:\s
                \t\tConsole: %d\t\tFile: %d

                """, CONS_NUM, FILE_NUM);

        int path;
        SaveData SD = new SaveData(true, 0);
        do{
            System.out.print("Your choice: ");
            path = pathCheck(SD, in);
        } while (SD.isCorrect());

        return path;
    }


    // block of condition check
    static boolean isArrIncreasing(double[] arrOfNumb) {
        for (int i = 1; i < arrOfNumb.length; i++) {
            if (!(arrOfNumb[i] <= arrOfNumb[i - 1])) {
                return false;
            }
        }

        return true;
    }


    static int resultOfArrChecking(boolean isIncrease){
        if (isIncrease)
            return 1;
        else
            return 0;
    }


    /// work in console
    static int inputArrSize(Scanner in){
        boolean isIncorrect = true;
        int arrSize = 0;

        do {
            System.out.print("How much number in massive: ");
            try {
                arrSize = Integer.parseInt(in.nextLine());
            } catch(Exception error){
                System.err.print("Invalid numeric input. Try again.\n");
            }
            if (arrSize < MIN_ARR_SIZE)
                System.err.printf("Min num is %d. Try again.\n", MIN_ARR_SIZE);
            else
                isIncorrect = false;
        } while (isIncorrect);

        return arrSize;
    }


    static double enteringTheCurrentNumber(int i, Scanner in){
        boolean isIncorrect = true;
        double currentNum = 0.0;

        do{
            System.out.printf("Write your %d number: ", i + 1);
            try{
                currentNum = Double.parseDouble(in.nextLine());
                isIncorrect = false;
            } catch(Exception error){
                System.err.print("Invalid numeric input. Try again.\n");
            }
        } while (isIncorrect);

        return currentNum;
    }


    static void inputArr(double[] arrOfNumb, int arrSize, Scanner in){
        for (int i = 0; i < arrSize; i++){
            arrOfNumb[i] = enteringTheCurrentNumber(i, in);
        }
    }


    static double[] viaConsole(Scanner in){
        SaveData SD = new SaveData(true, 0);
        SD.setArrSize(inputArrSize(in));
        double[] arrOfNumb = new double[SD.getArrSize()];

        inputArr(arrOfNumb, SD.getArrSize(), in);
        //boolean isIncreasing = isArrIncreasing(arrOfNumb);

        return arrOfNumb;//resultOfArrChecking(isIncreasing);
    }


    /// work with file
    static void fileRestriction(){
        System.out.print("""
                \nRules for storing information in a file:
                \t1.  The first line contains an integer:
                \t\tthe number of array elements;
                \t2.  The second line is real number
                \t\tentered separated by spaces.
                
                """);
    }


    static void wayCondition(String way, SaveData SD){
        if (way.length() > 4) {
            String bufstr = way.substring(way.length() - 4);
            if (bufstr.equals(".txt"))
                SD.setCorrect(false);
            else
                System.err.print("Write .txt file.\n");
        }
        else
            System.err.print("The path is too short.\n");
    }


    static String inputWayToTheFile(Scanner in){
        String way;
        SaveData SD = new SaveData(true, 0);

        System.out.print("Write way to your file: ");
        do{
            way = in.nextLine();

            wayCondition(way, SD);
        } while (SD.isCorrect());

        return way;
    }


    static boolean isFileIntegrity(String fileWay)
    {
        boolean isIntegrity = false;

        File file = new File(fileWay);
        if (file.canRead())
            isIntegrity = true;

        return isIntegrity;
    }

    static double[] resultOfReading(boolean isCorrect, double[] arrOfNumb)
    {
        if (isCorrect)
            return arrOfNumb;
        else {
            System.err.print("ERROR in file.");

            return null;
        }
    }


    static double[] isCorrectInputFromFile(Scanner in, SaveData SD){
        boolean isCorrect = true;
        double[] arrOfNumb = new double[0];
        if (!in.hasNextInt())
            isCorrect = false;
        else {
            int arrSize = in.nextInt();
            SD.setArrSize(arrSize);
            if (arrSize < MIN_ARR_SIZE)
                isCorrect = false;

            arrOfNumb = new double[arrSize];
            for (int i = 0; i < arrSize; i++) {
                if (!in.hasNextDouble())
                    isCorrect = false;
                try{
                    arrOfNumb[i] = in.nextDouble();
                } catch(Exception error) {
                    isCorrect = false;
                }
            }

            if (in.hasNext())
                isCorrect = false;
        }
        return resultOfReading (isCorrect, arrOfNumb);
    }


    static double[] isReadingCorrect(String fileWay, SaveData SD)  {
        try {
            Scanner fileScanner = new Scanner(new File(fileWay));
            return isCorrectInputFromFile(fileScanner, SD);
        } catch(FileNotFoundException error) {
            System.err.print("File not found.\n");
            return null;
        }
    }

    static double[] workWithFile(String fileWay) {
        SaveData SD = new SaveData(true, 0);

        return isReadingCorrect(fileWay, SD);
    }


    static double[] workWithIntegrityResult(boolean isIntegrity, String fileWay) {
        if (isIntegrity)
            return workWithFile(fileWay);
        else
        {
            System.err.print("Bad File.");

            return null;
        }
    }


    static double[] viaFile(Scanner in){
        fileRestriction();

        String fileWay = inputWayToTheFile(in);
        boolean isIntegrity = isFileIntegrity(fileWay);

        return workWithIntegrityResult(isIntegrity, fileWay);
    }


    /// output console
    static void outputViaConsole(int result){
        if (result == 1)
            System.out.print("Uncreased.");
        else
            System.out.print("Increasing.");
    }


    // output file
    static String fileCorrectOutput(int result){
        if (result == 1)
            return "\nUncreased.";
        else
            return "\nIncrease.";
    }


    static void outputViaFile(int result, Scanner in)
    {
        String fileWay = inputWayToTheFile(in);
        File file = new File(fileWay);
        StringBuilder builder;
        if (file.canWrite())
        {
            try{
                FileWriter writer = new FileWriter(fileWay);
                builder = new StringBuilder();
                builder.append(fileCorrectOutput(result));
                writer.write(builder.toString());
                writer.close();
                System.out.print("Check your file.");
            } catch (Exception error){
                System.err.print("Can't write in this file.");
            }
        }
        else
            System.err.print("\nBad output file.");
    }


    /// output
    static void output(int result, Scanner in)
    {
        if (result != -1)
        {
            System.out.print("\n\nYou need to choose where to output the result.\n");
            int option = choosingAPath(in);

            if (option == FILE_NUM) {
                outputViaFile(result, in);
            } else {
                outputViaConsole(result);
            }
        }
    }


    public static void main(String[] args){
        Scanner in = new Scanner(System.in);

        printStatement();

        int option = choosingAPath(in);
        double[] arrOfNumb = option == FILE_NUM ? viaFile(in) : viaConsole(in);

        if (arrOfNumb != null) {
            boolean isIncreasing = isArrIncreasing(arrOfNumb);
            int result = resultOfArrChecking(isIncreasing);
            output(result, in);
        }

        in.close();
    }
}